rllib/core/torch/torch_sarl_trainer.py

```diff
@@ -0,0 +1,172 @@
+    def compute_loss(batch, fwd_out, device, **kwargs):
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_grads_and_apply_if_needed(
+        batch, fwd_out, loss_out, rl_module, optimizer, device, **kwargs
+    ):
+        if not isinstance(optimizer, torch.optim.Optimizer):
+            for _opt in optimizer:
+                _opt.zero_grad()
+        else:
+            optimizer.zero_grad()
+        loss = loss_out["total_loss"]
+        loss.backward()
+        if not isinstance(optimizer, torch.optim.Optimizer):
+            for _opt in optimizer:
+                _opt.step()
+        else:
+            optimizer.step()
+
+        return {}
+
+    @staticmethod
+    def compile_results(
+        batch,
+        fwd_out,
+        loss_out,
+        compute_grads_and_apply_if_needed_info_dict,
+        rl_module,
+        **kwargs
+    ):
+        raise NotImplementedError
+
+    def get_weights(self):
+        # We need to get the weights from the trainer to the workers
+        return self.curr_weights
+
+    def _make_ray_train_trainer(self, scaling_config, init_rl_module_fn, module_config):
+        init_rl_module_fn = init_rl_module_fn or self.init_rl_module
+        queues = [Queue(maxsize=1) for _ in range(scaling_config.num_workers)]
+        backend_config = TorchConfig()
+        config = {
+            "rl_module_config": module_config,
+            "init_rl_module_fn": init_rl_module_fn,
+            "init_optimizer_fn": self.init_optimizer,
+            "compute_grads_and_apply_if_needed_fn": (
+                self.compute_grads_and_apply_if_needed
+            ),
+            "compute_loss_fn": self.compute_loss,
+            "compile_results_fn": self.compile_results,
+            "queues": queues,
+        }
+        train_loop_per_worker = construct_train_func(
+            self._training_func,
+            config,
+            fn_arg_name="train_loop_per_worker",
+            discard_returns=True,
+        )
+        backend_executor = BackendExecutor(
+            backend_config=backend_config,
+            num_workers=scaling_config.num_workers,
+            num_cpus_per_worker=scaling_config.num_cpus_per_worker,
+            num_gpus_per_worker=scaling_config.num_gpus_per_worker,
+            max_retries=0,
+        )
+        checkpoint_manager = CheckpointManager(checkpoint_strategy=None, run_dir=None)
+        # Start the remote actors.
+        backend_executor.start(initialization_hook=None)
+        empty_spec = RayDatasetSpec(dataset_or_dict=None)
+        training_iterator = TrainingIterator(
+            backend_executor=backend_executor,
+            backend_config=backend_config,
+            train_func=train_loop_per_worker,
+            dataset_spec=empty_spec,
+            checkpoint_manager=checkpoint_manager,
+            checkpoint=None,
+            checkpoint_strategy=None,
+        )
+        return queues, training_iterator
+
+    @staticmethod
+    def _training_func(config):
```

Comment:
This is an extremely difficult code to understand. There is a lot of information leakage between function calls that makes understanding it very difficult. We don't wanna create another version of what we already have man :)

Reply:
@kouroshHakha:
```
There is a lot of information leakage between function calls that makes understanding it very difficult.
```
This function only gets called once on each train worker, when training starts. The config arg with magic str-keys in it is not pretty, but enforced by Ray Train design, unfortunately.

```
We don't wanna create another version of what we already have man
```
We also don't want to reinvent the wheel by NOT using Ray Train and doing our own DDP/multi-GPU/etc.. Gotta work with what we have here. The upside of all this is that now we'll have a generic design that allows us to scale (on the learner end!), across all algos and DL-frameworks.
That being said, yes, we should try to simplify (and make readable) this as much as possible.